server/src/main/java/org/elasticsearch/index/mapper/KeyedJsonAtomicFieldData.java

```diff
@@ -0,0 +1,139 @@
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+
+public class KeyedJsonAtomicFieldData implements AtomicOrdinalsFieldData {
+
+    private final String key;
+    private final AtomicOrdinalsFieldData delegate;
+
+    KeyedJsonAtomicFieldData(String key,
+                             AtomicOrdinalsFieldData delegate) {
+        this.key = key;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public long ramBytesUsed() {
+        return delegate.ramBytesUsed();
+    }
+
+    @Override
+    public Collection<Accountable> getChildResources() {
+        return Collections.emptyList();
+    }
+
+    @Override
+    public SortedSetDocValues getOrdinalsValues() {
+        SortedSetDocValues values = delegate.getOrdinalsValues();
+        return new KeyedJsonDocValues(key, values);
+    }
+
+    @Override
+    public void close() {
+        delegate.close();
+    }
+
+    @Override
+    public ScriptDocValues<?> getScriptValues() {
+        return AbstractAtomicOrdinalsFieldData.DEFAULT_SCRIPT_FUNCTION
+            .apply(getOrdinalsValues());
+    }
+
+    @Override
+    public SortedBinaryDocValues getBytesValues() {
+        return FieldData.toString(getOrdinalsValues());
+    }
+
+    private static class KeyedJsonDocValues extends AbstractSortedSetDocValues {
+
+        private final BytesRef prefix;
+        private final SortedSetDocValues delegate;
+
+        KeyedJsonDocValues(String key, SortedSetDocValues delegate) {
+            this.prefix = new BytesRef(JsonFieldParser.createKeyedValue(key, ""));
+            this.delegate = delegate;
+        }
+
+        @Override
+        public long getValueCount() {
+            return delegate.getValueCount();
+        }
+
+        @Override
+        public BytesRef lookupOrd(long ord) throws IOException {
+            BytesRef keyedValue = delegate.lookupOrd(ord);
+            int valueLength = keyedValue.length - prefix.length;
+            return new BytesRef(keyedValue.bytes, prefix.length, valueLength);
+        }
+
+        @Override
+        public long nextOrd() throws IOException {
+            for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                if (accepted(ord)) {
+                    return ord;
+                }
+            }
+            return NO_MORE_ORDS;
+        }
+
+        @Override
+        public boolean advanceExact(int target) throws IOException {
+            if (delegate.advanceExact(target)) {
+                for (long ord = delegate.nextOrd(); ord != NO_MORE_ORDS; ord = delegate.nextOrd()) {
+                    if (accepted(ord)) {
+                        boolean advanced = delegate.advanceExact(target);
+                        assert advanced;
+                        return true;
+                    }
+                }
+            }
+            return false;
+        }
+
+        private boolean accepted(long ord) throws IOException {
```

Comment:
Since ordinals are sorted it should be possible to compute the range of ordinals that belongs to a single prefix (field). You can perform a binary search to find the first and last ordinals that contains the field prefix and use this information to remove the need to call `lookupOrd` on each value.

Reply:
Nice, will try that out!

Reply:
I looked into this more closely, and wanted to check I understand your suggestion: when creating a new `KeyedJsonAtomicFieldData`, we can do a one-time calculation to figure out the relevant range of ordinals. If the set of documents considered during the search is very restricted, this might be slightly more expensive than the current approach, but in general it should cut down a lot on lookups/ bytes comparisons.

Reply:
That's the idea yes, we'll need to recompute the min/max for each query (since we want to cache the fielddata only once for all fields) but since it's a binary search it shouldn't be too expensive.